lib/bootstrapper.js

```diff
@@ -0,0 +1,202 @@
+
+  var appConfigRootDir = options.appConfigRootDir || appRootDir;
+
+  options.rootDir = appConfigRootDir;
+  options.env = env;
+  this.options = options;
+
+  this.phases = options.phases || builtinPhases;
+  this.builtinPlugins = options.plugins || builtinPlugins;
+  assert(Array.isArray(this.phases), 'Invalid phases: ' + this.phases);
+  assert(Array.isArray(this.plugins), 'Invalid plugins: ' +
+    this.builtinPlugins);
+
+  var self = this;
+  self.builtinPlugins.forEach(function(p) {
+    var factory = require('./plugins/' + p);
+    self.use('/boot/' + p, factory(options));
+  });
+
+  try {
+    loadAndRegisterPlugins(self, options);
+  } catch (err) {
+    debug(err);
+  }
+}
+
+/**
+ * Register a handler to a given path
+ * @param {String} path
+ * @param {Function} handler
+ */
+Bootstrapper.prototype.use = function(path, handler) {
+  var plugin = {
+    path: path,
+    handler: handler,
+  };
+  this.plugins.push(plugin);
+};
+
+/**
+ * Get a list of plugins for the given path
+ * @param {String} path
+ * @returns {*}
+ */
+Bootstrapper.prototype.getPlugins = function(path) {
+  if (path[path.length - 1] !== '/') {
+    path = path + '/';
+  }
+  return this.plugins.filter(function(p) {
+    return p.path.indexOf(path) === 0;
+  });
+};
+
+/**
+ * Get a list of extensions for the given path
+ * @param {String} path
+ * @returns {*}
+ */
+Bootstrapper.prototype.getExtensions = function(path) {
+  if (path[path.length - 1] !== '/') {
+    path = path + '/';
+  }
+  return this.plugins.filter(function(p) {
+    if (p.path.indexOf(path) === -1) return false;
+    var name = p.path.substring(path.length);
+    return name && name.indexOf('/') === -1;
+  });
+};
+
+/**
+ * Add more phases. The order of phases is decided by the sequence of phase
+ * names
+ * @param {String[]} phases An array of phase names
+ * @returns {String[]} New list of phases
+ */
+Bootstrapper.prototype.addPhases = function(phases) {
+  this.phases = utils.mergePhaseNameLists(this.phases, phases || []);
+  return this.phases;
+};
+
+/**
+ * Invoke the plugins phase by phase with the given context
+ * @param {Object} context Context object
+ * @param {Function} done Callback function. If not provided, a promise will be
+ * returned
+ * @returns {*}
+ */
+Bootstrapper.prototype.run = function(context, done) {
+  if (!done) {
+    done = createPromiseCallback();
+  }
+  var options = this.options;
+  var appRootDir = options.appRootDir = options.appRootDir || process.cwd();
+  var env = options.env || process.env.NODE_ENV || 'development';
+
+  var appConfigRootDir = options.appConfigRootDir || appRootDir;
+
+  options.rootDir = appConfigRootDir;
+  options.env = env;
+
+  context = context || {};
+
+  var phases = context.phases || this.phases;
+  var bootPlugins = this.getExtensions('/boot');
+  async.eachSeries(phases, function(phase, cb1) {
+    debug('Phase %s', phase);
+    async.eachSeries(bootPlugins, function(plugin, cb2) {
```

Comment:
Having to call variables like `cb1` and `cb2` is a code smell. You should use more descriptive names, e.g. `nextPhase` and `nextPlugin`.

I think it will be even better to refactor this code and move inner anonymous functions out of this method into regular functions/methods.